File: 20_valid_parentheses/isValid.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        unordered_map<char, char> map;
        map[')'] = '(';
        map['}'] = '{';
        map[']'] = '[';

        for (auto i : s) {
            if (map.find(i) != map.end()) {
                if (!stack.empty() && stack.top() == map[i]) {
                    stack.pop();
                } else {
                    return false;
                }
            } else {
                stack.push(i);
            }
        }

        if (!stack.empty())
            return false;

        return true;
    }
};

File: 20_valid_parentheses/isValid.ts
function isValid(s: string): boolean {
  const characters = ["{", "(", "["];

  const stack = [];

  for (let i of s) {
    if (characters.includes(i)) {
      stack.push(i);
    } else {
      if (stack.length > 0) {
        let last = stack.pop();
        if (last === "{" && i !== "}") return false;
        if (last === "(" && i !== ")") return false;
        if (last === "[" && i !== "]") return false;
      } else {
        return false;
      }
    }
  }

  if (stack.length > 0) return false;

  return true;
}
